### ISSUE TITLE
diag(p3): class-balanced BC-fit — discriminate capacity gap from class-imbalance underfitting (Phase 1.5)

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/bucket-brigade#279]
## Hypothesis being tested

PR #276 (closes #272) found BC at default config fails the discriminating sub-task: house-on-WORK head collapses to chance (0.354 ≈ 1/3 over 3 owned houses), and all heads' headline accuracy *exactly* matches the class prior P(REST)=0.9435. **Two hypotheses are equally consistent**:

1. **Architecture capacity gap** — the network can't represent the burning-owned argmax under any training regime
2. **Class-imbalance underfitting** — with 19:1 REST:WORK ratio, gradients on the discriminating sub-task are diluted; the network is trapped at the class-prior local optimum despite having capacity

Judge on #276 explicitly recommended this experiment as the cheapest discriminator (~3-line change).

## Protocol

1. Reuse `experiments/p3_specialization/bc_fit_only.py` from PR #276
2. Apply **class-balanced loss weighting** to the mode head — sample weight inversely proportional to class frequency, so each REST and WORK contributes equally to the loss
3. Optionally: **oversample** WORK rows during training to further amplify their gradient signal
4. Train for the same 10 epochs (extend if loss still decreasing — PR #276's loss was actively decreasing at epoch 10, *underfitting* not overfitting)
5. Compute the same per-head accuracy breakdown, especially **house head on mode=WORK subset**

## Verdict logic

| house-on-WORK accuracy | Interpretation | Next step |
|---|---|---|
| **≥ 0.90** | **Class-imbalance was the failure mode.** Architecture has capacity; PPO's failure is reward-distribution / advantage-weighting. | File: combine class-balancing + PPO (advantage reweighting?) |
| **0.50 - 0.90** | Partial. Class-balancing helps but doesn't fully solve. Both effects contribute. | Layer on hidden_size sweep (#279) to test residual capacity |
| **≤ 0.50** | Class-balancing didn't help. Confirmed capacity gap. | Escalate to hidden_size sweep + Transformer (#279, #280) |

## Cost

~30 min on CPU. Same demo gen + supervised loop as #272.

## Acceptance criteria

- [ ] `experiments/p3_specialization/bc_fit_balanced.py` or extension of `bc_fit_only.py` with `--class-balanced` flag
- [ ] Per-head accuracy report + WORK-subset breakdown
- [ ] Verdict applied per the ladder above
- [ ] Notebook entry comparing to PR #276's unbalanced baseline
- [ ] If verdict 1: project memory updated; further capacity work deprioritized
- [ ] If verdict 3: open follow-up #279 (hidden_size sweep)

## References

- PR #276 (the unbalanced baseline this builds on)
- Judge of #276 — methodological correction that motivated this issue
- docs/ppo-help-brief.md

**Comment by rjwalters** [comment]
Curating this issue per user request

**Comment by rjwalters** [comment]
## Curator Enhancement

The issue's hypothesis and verdict table are crisp. This comment fills in concrete file paths, line refs, and three specific implementation options so a Builder can start immediately. It also adds the missing Dependencies, Affected Files, and Test Plan sections required by the Builder quality gate.

### Codebase context (file:line refs against current `main`)

**BC-fit training loop & loss** — `experiments/p3_specialization/bc_fit_only.py:132-249`:
- The per-batch loss is the sum of unweighted `F.cross_entropy(logits[k], yb[:, k])` across `k ∈ {house, mode, signal}` (lines 181-182).
- Eval is identical (no weighting) at lines 192-203.
- The script already computes a `house_acc_on_work_subset` after training (lines 227-237) — that is the headline discriminating number this experiment needs to move.

**Oracle data source** — `bucket_brigade/baselines/specialist.py:55-118` (`specialist_action`) and `:119-138` (`specialist_action_joint`). Deterministic: `house = lowest_burning_owned (WORK) or lowest_owned (REST)`, `mode = WORK iff any owned burning`, `signal = mode`.

**Empirical action-class distribution** (from `experiments/p3_specialization/bc_fit_only_result.json`):
- `mode_counts: [9505 REST, 495 WORK]` → **19.2:1 imbalance**, work_frac = 0.0495
- `signal_counts` identical (specialist sets signal=mode)
- `house_counts`: agents 0-3 dominate (~2400 each = `lowest_owned` for REST rows); houses 4-9 are sparse (~50 each = burning-owned WORK rows)
- This 19.2:1 ratio is the exact lever the experiment targets.

**Network** — `bucket_brigade.training.networks.PolicyNetwork` (instantiated at `bc_fit_only.py:164-166`): 3 independent categorical heads, ~7.9k params, `hidden_size=64`.

### Implementation options (pick one or combine)

The verdict-discriminating result is the **house-on-WORK subset accuracy**. There are three canonical ways to lift its gradient signal; each is ≤10 lines of change inside `train_bc`:

| Option | Mechanism | Code shape | Pros / Cons |
|---|---|---|---|
| **A. Weighted CE on mode/signal heads** | `F.cross_entropy(logits[k], yb[:, k], weight=class_weight_k)` with `class_weight_k = N / (2 * bincount(yb[:,k]))` | One-line per head; class weights computed once from `labels` | Direct fix for 2-way heads. Doesn't touch the house head's gradient distribution. |
| **B. Oversample WORK rows** | Replace `torch.randperm` minibatch builder (line 175) with a `WeightedRandomSampler` weighting WORK rows ~19× higher | ~10 lines: build per-row weights from `labels[:,1]`, use `torch.utils.data.WeightedRandomSampler` | Lifts gradient on all 3 heads simultaneously (because WORK rows dominate each batch). The strongest single intervention for the house-on-WORK metric. |
| **C. Subsample REST rows** | Downsample REST rows to ~equal count before training (drops ~9000 rows from a 10k set) | ~5 lines pre-train filter | Cheapest, but discards data and biases eval split if applied naïvely. Apply only to train split. |

**Recommended default**: combine **A + B** behind a single `--class-balanced` flag. Option B is the primary mechanism (lifts house-head gradient signal on WORK rows); option A backs it up on the mode/signal heads. Option C is a fallback if B's variance is too high.

### Success criterion (operationalized for this issue)

Headline number: **`house_acc_on_work_subset`** (the only non-trivial discriminating metric; the other two heads are 2-way and trivially saturated at ~94.4% via class-prior prediction).

| `house_acc_on_work_subset` | Verdict | Project-memory update |
|---|---|---|
| ≥ 0.90 | **CLASS_IMBALANCE confirmed.** PR #276's INDUCTIVE_BIAS_GAP verdict was a class-imbalance artifact. Architecture has capacity. | Update `project_thesis_misaligned_gradients.md`: reframe PPO failure as advantage-weighting, not capacity. Deprioritize #280. |
| 0.50 - 0.90 | **PARTIAL.** Class-balancing helps but doesn't fully solve. | Run #280 (hidden_size sweep) to test residual capacity bound. |
| ≤ 0.50 | **CAPACITY confirmed.** Class-balancing alone doesn't help; the issue is representational. | Escalate #280 immediately; consider Transformer policy. |

Secondary numbers to report (for reproducibility and to detect collapse to the opposite class prior):
- Eval loss at final epoch (and trajectory)
- Per-head accuracy: house / mode / signal
- Joint accuracy
- WORK-subset row count in eval split (~113 in the unbalanced baseline; oversampling does NOT change eval split composition)
- Number of WORK rows seen by gradient per epoch (with oversampling, should be ~5-10× higher than baseline)

### Comparison protocol vs. #276 baseline

Use the same `--seed 0`, same `num_steps=2500`, same `--epochs 10` (and a longer `--epochs 30` run to match PR #276's plateau observation). Report a side-by-side table:

```
                          #276 (unbalanced)   #279 (balanced)
eval_loss (epoch 10)      0.8058              ???
eval_loss (epoch 30)      ~0.59               ???
house_acc                 0.9635              ???
house_acc_on_work_subset  0.354               ???   ← THE NUMBER
mode_acc                  0.9435              ???
joint_acc                 0.???               ???
```

The "underfitting at epoch 10" observation from the Judge of #276 is important: do not stop early. Train at least to epoch 30 (eval loss should still be falling), and ideally to convergence.

## Dependencies

- [ ] #278: BC-init + PPO continuation (basin-trap probe) — merged via PR (verify in issue body of #279 — the issue links the wrong number; sister issue should be #272/PR #276 per the body, since #278 is the BC-init basin-trap probe, distinct lineage). **Practical gate**: PR #276 (closes #272) is **merged** as of `main`, so the prerequisite `bc_fit_only.py` and result JSON are present. This issue can proceed.

Note on issue-number cross-references: the body of #279 references `#279` self-referentially as the hidden_size follow-up — that follow-up is actually filed as **#280**. The Builder should treat #280 as the hidden_size escalation.

## Affected Files

- `experiments/p3_specialization/bc_fit_only.py` — extend with `--class-balanced` flag, weighted-CE construction (lines ~181-182), and `WeightedRandomSampler` minibatch path (replacing `torch.randperm` at line 175). Keep default behavior bit-identical when flag is off.
- `experiments/p3_specialization/bc_fit_balanced_result.json` (new, or write to a flag-conditional path) — separate output file so the unbalanced #276 baseline JSON is not overwritten.
- `experiments/p3_specialization/bc_fit_only_result.json` — read-only baseline reference (do not modify).
- `bucket_brigade/baselines/specialist.py` — read-only; label source.
- `bucket_brigade/training/networks.py` — read-only; `PolicyNetwork` target.
- `docs/ppo-help-brief.md` — append a paragraph with the verdict outcome (small documentation update).
- `~/.claude/projects/-Users-rwalters-GitHub-bucket-brigade/memory/project_thesis_misaligned_gradients.md` — update IF verdict is CLASS_IMBALANCE (reframes the PPO failure thesis).

Out of scope (explicit non-goals):
- `hidden_size` sweep — that is #280.
- PPO continuation runs — out of scope; this issue is BC-only.
- Production training script changes — only the diagnostic script changes.

## Test Plan

- [ ] **Manual verification**: `uv run python experiments/p3_specialization/bc_fit_only.py --class-balanced --epochs 30 --seed 0 --out experiments/p3_specialization/bc_fit_balanced_result.json` runs to completion and prints the FINAL SUMMARY block with `house_acc_on_work_subset` value.
- [ ] **Backward-compatibility regression**: Running without `--class-balanced` reproduces the PR #276 numbers to ~3 decimal places (`house_acc_on_work_subset ≈ 0.354`, `mode_acc ≈ 0.9435`). This guards against accidentally altering the unbalanced code path.
- [ ] **Class-weight arithmetic check**: With `mode_counts = [9505, 495]`, weighted-CE weight vector should be `[0.526, 10.10]` (or equivalent normalization). Print and assert this once at training start.
- [ ] **Oversampler check**: With `WeightedRandomSampler`, per-epoch WORK-row exposure (`sum of yb[:,1]`) should be ~50% of total batch rows, not ~5%. Log the per-epoch WORK fraction.
- [ ] **Reproducibility**: Two runs with `--seed 0 --class-balanced` produce identical final metrics to 3 decimals.
- [ ] **Verdict logged**: Update `docs/ppo-help-brief.md` and/or `experiments/p3_specialization/` notebook entry with the verdict per the table above. Cross-reference to the #276 baseline JSON for side-by-side comparison.
- [ ] **Automated tests**: N/A (research diagnostic; no production code paths change).
- [ ] **Edge case — no WORK rows in eval split**: Unlikely at 4.95% with n=2000 eval (~99 WORK rows expected), but guard against `work_count == 0` is already in `bc_fit_only.py:236-237`.

## References

- PR #276 (closes #272) — the unbalanced baseline this builds on. Judge's review (top comment of PR #276) is the methodological correction that motivated this issue.
- Sister issue #280 — hidden_size sweep escalation (run only if this issue's verdict is PARTIAL or CAPACITY).
- Sister issue #278 — BC-init + PPO continuation (PR #270, merged); not a hard dependency for this BC-only diagnostic.
- `docs/ppo-help-brief.md` — broader Phase-1 trio context.
- Project memory: `project_thesis_misaligned_gradients.md`, `project_ppo_failure_mode.md` — both updated based on outcome.

**Comment by rjwalters** [comment]
Curation complete. Added file:line refs (bc_fit_only.py:175,181-182,227-237 and specialist.py:55-138), three concrete implementation options (weighted CE / WeightedRandomSampler / subsample), operationalized success criterion on house_acc_on_work_subset with verdict ladder, Dependencies / Affected Files / Test Plan sections. Duplicate check clear (only self-match). Awaiting human approval via loom:issue.
